mev_inspect/block.py

```diff
@@ -28,13 +29,12 @@ async def create_from_block_number(
     base_provider,
     w3: Web3,
     block_number: int,
-    trace_db_session: Optional[orm.Session],
 ) -> Block:
     block: Optional[Block] = None
 
-    if trace_db_session is not None:
-        block = _find_block(trace_db_session, block_number)
-
+    if get_trace_session() is not None:
+        async with get_trace_session() as session:  # type: ignore
+            block = await _find_block(session, block_number)
```

Comment:
might be what you meant by tracedb issues, but in the inspect-many case we'll open / close a new connection to the DB on each iteration of this loop. would be ideal to share across

Reply:
Updated to have a db session created once per task, similarly to how you structured your PR. Warnings still persist, but it is indeed a nicer set up. Let me know what you think

Reply:
It's better, but I think still better to have it at the even higher level

SQLAlchmey recommends having a single session shared across the duration of the script run v. creating at the task level - ​https://docs.sqlalchemy.org/en/14/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it

Right now if I call two Inspector functions, it will create a separate session for each

Since we're using the NullPool especially, I believe that means we'll experience the overhead of connecting to the database / closing connection on each call v. getting to reuse the connection across the script.

Reply:
Totally fair -- I've updated to follow the suggestion. The warnings still show up, but everything works -- please feel free to test it locally! It also seems to be right pattern according to the docs examples regarding usage of async sqlachemy with asyncio.gather: https://docs.sqlalchemy.org/en/14/_modules/examples/asyncio/gather_orm_statements.html

> Note that this pattern loses all transactional safety and is also not
necessarily any more performant than using a single Session, as it adds
significant CPU-bound work both to maintain more database connections
and sessions, as well as within the merging of results from external sessions
into one.

Reply:
Wooo awesome! I'll give a try today!

Wild that they're like "here's how you do it - btw this loses all transactional safety and isn't better"

Ideally we want literally the same thing as the non-async version only without breaking lol. Single threaded and blocking.

Reply:
I wonder if we'll have to switch off the NullPool.

Pretty sure NullPool means exactly one connection at a time. It sounds like from this they create a separate connection for each async usage, which with NullPool might mean opening and closing the connection each time we jump between async functions.

Reply:
From your comment it seems to me that actually, maybe the first iteration here should use the pattern of just running everything in a transaction block in sync mode with

```
 await session.run_sync(run_queries)
```

See [example](https://docs.sqlalchemy.org/en/14/_modules/examples/asyncio/greenlet_orm.html)

I'll wait for your feedback on current usage, but I'll do experimentation with this. 

An your notes, I believe you're correct. Each async task would indeed use a new connection, so we might want to switch out of NullPull -- will investigate what are the possibilities there

Reply:
> The function will be passed a synchronous-style Session object and the function can use traditional ORM patterns.

Wow that's pretty sweet - would mean we don't need to change the current crud logic right?

Reply:
On this https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-multiple-asyncio-event-loops
it says
> An application that makes use of multiple event loops, for example by combining asyncio with multithreading, should not share the same AsyncEngine with different event loops when using the default pool implementation.
which makes me think it does actually re-use the same connections across async

I'll run a test on this branch and report back

Reply:
Trying locally, I did see it create multiple connections. And then it seemed like everything locked up.

To replicate, I did
```
./mev inspect-many 13115099 13115299
```

Then was checking DB connections by running
```
select pid as process_id,
       usename as username,
       datname as database_name,
       client_addr as client_address,
       application_name,
       backend_start,
       state,
       state_change
from pg_stat_activity;
```

which looked like this for a bit
```
 process_id | username | client_address | application_name |        state
------------+----------+----------------+------------------+---------------------
        997 |          |                |                  |
        999 | postgres |                |                  |
       1399 | postgres | 10.244.0.55    | psql             | active
       2523 | postgres | 10.244.0.59    |                  | idle
       2525 | postgres | 10.244.0.59    |                  | idle in transaction
       2527 | postgres | 10.244.0.59    |                  | idle in transaction
       2529 | postgres | 10.244.0.59    |                  | active
       2531 | postgres | 10.244.0.59    |                  | idle in transaction
```

then froze like this
```
 process_id | username | client_address | application_name | state
------------+----------+----------------+------------------+--------
        997 |          |                |                  |
        999 | postgres |                |                  |
       1399 | postgres | 10.244.0.55    | psql             | active
       2523 | postgres | 10.244.0.59    |                  | active
       2525 | postgres | 10.244.0.59    |                  | active
       2527 | postgres | 10.244.0.59    |                  | active
       2529 | postgres | 10.244.0.59    |                  | active
       2531 | postgres | 10.244.0.59    |                  | active
```